File: Top-Interview-Questions/Array-and-Strings-Medium-Collection/163-missing-ranges.js
/**
 * @param {number[]} nums
 * @param {number} lower
 * @param {number} upper
 * @return {string[]}
 * Runtime: 64 ms, faster than 67.02% of JavaScript online submissions for Missing Ranges.
 * Memory Usage: 33.3 MB, less than 65.62% of JavaScript online submissions for Missing Ranges.
 * TC: O(N) where N is number of element in nums
 * SC: O(1)
 */
function findMissingRanges(nums, lower, upper) {
    const res = [];
    nums = [lower - 1, ...nums, upper + 1];

    for (let i = 1; i < nums.length; i++) {
        const diff = nums[i] - nums[i - 1];

        if (diff === 2) {
            res.push(`${nums[i - 1] + 1}`);
        } else if (diff > 2) {
            res.push(`${nums[i - 1] + 1}->${nums[i] - 1}`);
        }
    }

    return res;
}

File: Top-Interview-Questions/Array-and-Strings-Medium-Collection/163-missing-ranges.py
class Solution:
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[str]:
        # str()+"->"+str()
        # append
        # range varies from lower+1 to n-1
        check = None
        result = []
        for num in nums+[upper+1]:
            if num == check:
                continue
            if num != lower:
                if num == lower+1:
                    result.append(str(lower))
                else:
                    result.append(str(lower)+"->"+str(num-1))
                    
            lower = num + 1
            check = num
        
        return result

        # TC: O(n)
        # SC: O(n)
        # Runtime: 36 ms, faster than 28.63%
        # Memory Usage: 14 MB, less than 15.48%